atomicapp/providers/lib/kubeshift/openshift.py

```diff
@@ -0,0 +1,384 @@
+            src (str): Source path in container
+            dest (str): Destination path in host
+            update (bool): Update existing destination, if True
+        """
+        if os.path.exists(dest) and not update:
+            return
+        cleaned_image_name = Utils.sanitizeName(image)
+        pod_name = '{}-{}'.format(cleaned_image_name, Utils.getUniqueUUID())
+        container_name = cleaned_image_name
+
+        # Pull (if needed) image and bring up a container from it
+        # with 'sleep 3600' entrypoint, just to extract content from it
+        artifact = {
+            'apiVersion': 'v1',
+            'kind': 'Pod',
+            'metadata': {
+                'name': pod_name
+            },
+            'spec': {
+                'containers': [
+                    {
+                        'image': image,
+                        'command': [
+                            'sleep',
+                            '3600'
+                        ],
+                        'imagePullPolicy': 'IfNotPresent',
+                        'name': container_name
+                    }
+                ],
+                'restartPolicy': 'Always'
+            }
+        }
+
+        self.create(artifact, namespace)
+        try:
+            self._wait_till_pod_runs(namespace, pod_name, timeout=300)
+
+            # Archive content from the container and dump it to tmpfile
+            tmpfile = '/tmp/atomicapp-{pod}.tar.gz'.format(pod=pod_name)
+
+            self._execute(
+                namespace, pod_name, container_name,
+                'tar -cz --directory {} ./'.format('/' + src),
+                outfile=tmpfile
+            )
+        finally:
+            # Delete created pod
+            self.delete(artifact, namespace)
+
+        # Extract archive data
+        tar = tarfile.open(tmpfile, 'r:gz')
+        tar.extractall(dest)
+
+    def _execute(self, namespace, pod, container, command,
+                 outfile=None):
+        """
+        Execute a command in a container in an Openshift pod.
+
+        Args:
+            namespace (str): Namespace
+            pod (str): Pod name
+            container (str): Container name inside pod
+            command (str): Command to execute
+            outfile (str): Path to output file where results should be dumped
+
+        Returns:
+            Command output (str) or None in case results dumped to output file
+        """
+        args = {
+            'token': self.api.token,
+            'namespace': namespace,
+            'pod': pod,
+            'container': container,
+            'command': ''.join(['command={}&'.format(word) for word in command.split()])
+        }
+        url = urljoin(
+            self.k8s_api,
+            'namespaces/{namespace}/pods/{pod}/exec?'
+            'access_token={token}&container={container}&'
+            '{command}stdout=1&stdin=0&tty=0'.format(**args))
+
+        self.api.websocket_request(url, outfile)
```

Comment:
you are missing the return here if the user didn't pass in `outfile`. here is what it was:

```
if not outfile:
    return ''.join(results)
```